incubator/hnc/controllers/hierarchy_controller.go

```diff
@@ -47,76 +49,144 @@ type HierarchyReconciler struct {
+		nsnnm := types.NamespacedName{Name: nm}
+		if err := r.Get(ctx, nsnnm, ns); err != nil {
+			if !errors.IsNotFound(err) {
+				log.Error(err, "Couldn't read namespace", "ns", nm)
+				return err
+			}
+
+			// The namespace is gone. Remove all traces of the namespace from  our data structures.
+			log.Info("Namespace was deleted", "ns", nm)
+			return r.onDelete(ctx, log, nm)
+		}
+
+		// If the namespace is being deleted, we should still sync any changes that we might have missed
+		// (ie, maybe the parent used to be set?) but then don't update anything.
+		if !ns.GetDeletionTimestamp().IsZero() {
+			log.Info("Namespace is being deleted", "ns", nm)
+			update = false
+		}
+	}
+
+	// Sync the tree. If `ok` is false, that means that there are conditions on this namespace and we
+	// shouldn't continue to sync the objects.
+	if ok, err := r.updateTree(ctx, log, inst, update); !ok || err != nil {
+		return err
 	}
 
-	return ctrl.Result{}, r.update(ctx, log, inst, true)
+	if update {
+		// Update all the objects in this namespace. We have to do this at least *after* the tree is
+		// updated, because if we don't, we could incorrectly think we've propagated the wrong objects
+		// from our ancestors, or are propagating the wrong objects to our descendants.
+		return r.updateObjects(ctx, log, inst.ObjectMeta.Namespace)
+	}
+
+	return nil
 }
 
-// update performs three tasks: it syncs the Hierarchy singleton with the in-memory forest; it calls
-// itself on any affected namespaces if the hierarchy has changed, and it invokes all the
-// TypeReconcilers to propagate objects. It *always* results in one write to the apiserver per
-// instance, even if no changes are needed (we can fix this later).
-func (r *HierarchyReconciler) update(ctx context.Context, log logr.Logger, inst *tenancy.Hierarchy, updateObjects bool) error {
-	log.V(1).Info("Updating")
+// updateTree syncs the Hierarchy singleton with the in-memory forest (writing back to the apiserver
+// if necessary and requested) and calls itself on any affected namespaces if the hierarchy has
+// changed. It returns false if there are conditions on this namespace, which indicates that any
+// changes from this namespace should not be further propagated.
+//
+// TODO: store the conditions in the in-memory forest so that object propagation can be disabled if
+// there's a problem on the namespace.
+func (r *HierarchyReconciler) updateTree(ctx context.Context, log logr.Logger, inst *tenancy.Hierarchy, update bool) (bool, error) {
+	// Update the in-memory data structures
+	orig := inst.DeepCopy()
 	affected := r.syncWithForest(ctx, log, inst)
-	if err := r.Update(ctx, inst); err != nil {
-		log.Error(err, "while updating apiserver")
-		return err
+
+	// If appropriate, write back. We only write if there's been a *change* in the data structure, and
+	// if the caller wants us to update it (ie, if the namespace isn't being deleted).
+	if update && !reflect.DeepEqual(orig, inst) {
```

Comment:
In create new ns case, will reflect.DeepEqual(orig, inst) return true since syncWithForest will not change the content of inst object? Particularly,  I am not sure if adding inst.Status.Children = nil will alter the DeepEqual() result.

Reply:
I set inst.Status.Children=nil to make sure that DeepEqual worked correctly.  So yes, for new namespaces, DeepEqual does return true, and the `hier` object is not written back as a result (which is the intended behaviour).

Reply:
> > I am assuming you will add webhook to prevent hier object from being created by any user other than the controller.
> 
> No, users can create it if it doesn't exist - that's how you set the parent. However, the CRD validation schema will only allow a single name, so there can only be at most one per namespace.

Got you.